lib/map.h

```diff
@@ -17,42 +17,66 @@ limitations under the License.
 #ifndef LIB_MAP_H_
 #define LIB_MAP_H_
 
-#include <map>
+#include <optional>
 
-template <class K, class T, class V, class Comp, class Alloc>
-inline V get(const std::map<K, V, Comp, Alloc> &m, T key, V def = V()) {
+/// Given a map and a key, return the value corresponding to the key in the map,
+/// or a given default value if the key doesn't exist in the map.
+template <typename Map, typename Key>
+typename Map::mapped_type get(const Map &m, const Key &key) {
     auto it = m.find(key);
-    if (it != m.end()) return it->second;
-    return def;
+    return it != m.end() ? it->second : typename Map::mapped_type{};
+}
+template <class Map, typename Key = typename Map::key_type,
+          typename Value = typename Map::mapped_type>
+typename Map::mapped_type get(const Map &m, const Key &key, Value &&def) {
+    using M = typename Map::mapped_type;
+    auto it = m.find(key);
+    return it != m.end() ? it->second : static_cast<M>(static_cast<Value &&>(def));
```

Comment:
std::move instead of the inner static cast?

Reply:
Oh, no. Then lvalues will be moved. I would change it to `std::forward`, it will be more clear that we're doing a "universal reference" stuff here.

Reply:
Oh sorry about that, I did not realize it is universal reference.

Reply:
This part is always confusing...